### PR TITLE
feat: email allow list to protect from impersonation

### PR DESCRIPTION
**Pull request by remisultan**
## :id: Reference related issue. 

https://gravitee.atlassian.net/browse/AM-649

### CODE REVIEW

File: gravitee-am-service/src/main/java/io/gravitee/am/service/validators/flow/policy/SendEmailPolicyValidator.java [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+import static java.util.Optional.ofNullable;
+
+/**
+ * @author Rémi SULTAN (remi.sultan at graviteesource.com)
+ * @author GraviteeSource Team
+ */
+@Component
+public class SendEmailPolicyValidator implements PolicyValidator {
+
+    private final Logger LOGGER = LoggerFactory.getLogger(SendEmailPolicyValidator.class);
```
Comment (mariano-fernandez) [inline gravitee-am-service/src/main/java/io/gravitee/am/service/validators/flow/policy/SendEmailPolicyValidator.java:35]:
```suggestion
    private static final Logger LOGGER = LoggerFactory.getLogger(SendEmailPolicyValidator.class);
```

File: gravitee-am-service/src/main/java/io/gravitee/am/service/validators/flow/policy/SendEmailPolicyValidator.java [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+
+    @Override
+    public Boolean validate(Step element) {
+        if (POLICY_AM_SEND_EMAIL.equalsIgnoreCase(element.getPolicy())) {
+            try {
+                var tree = objectMapper.readTree(element.getConfiguration());
+                final String from = ofNullable(tree.get(FROM_KEY)).map(JsonNode::textValue).orElse(null);
+                return emailDomainValidator.validate(from);
+            } catch (Exception e) {
+                LOGGER.debug("An unexpected error has occurred", e);
```
Comment (mariano-fernandez) [inline gravitee-am-service/src/main/java/io/gravitee/am/service/validators/flow/policy/SendEmailPolicyValidator.java:54]: If something unexpected happens, we want to know about it. I would probably log this with `error` level

File: gravitee-am-service/src/test/java/io/gravitee/am/service/validators/FlowValidatorImplTest.java [hunk 3]
```diff
@@ -0,0 +1,149 @@
...
+        var observer = flowValidator.validate(flow).test();
+        observer.awaitDone(10, TimeUnit.SECONDS);
+        observer.assertComplete().assertNoErrors();
+
+        verify(emailPolicyValidator, times(2)).validate(any());
+    }
+
+    @ParameterizedTest
+    @MethodSource("params_that_must_invalidate_step")
+    void must_invalidate_step(String invalidStepName, int expectedInvocations) {
```
Comment (mariano-fernandez) [inline gravitee-am-service/src/test/java/io/gravitee/am/service/validators/FlowValidatorImplTest.java:69]: Can variables declaration be adjusted to have consistency? In the same method there are variables declared with `final`, some of them without it and other with `var`.
In general I prefer to use `var` and leave the compiler to do the work for me. But that's just personal preference 😄

File: gravitee-am-management-api/gravitee-am-management-api-service/src/main/java/io/gravitee/am/management/service/impl/EmailServiceImpl.java [hunk 4]
```diff
@@ -137,14 +138,13 @@ public Maybe<Email> send(Domain domain, Application client, io.gravitee.am.model
 
     private Maybe<Email> prepareAndSend(Domain domain, Application client, io.gravitee.am.model.Template template, User user) {
         // get raw email template
-        return getEmailTemplate(template, user).map(emailTemplate -> {
+        return getEmailTemplate(template, user).flatMap(emailTemplate -> {
```
Comment (leleueri) [inline gravitee-am-management-api/gravitee-am-management-api-service/src/main/java/io/gravitee/am/management/service/impl/EmailServiceImpl.java:141]: Question: Why are you switching to flatMap and so Initialize a Maybe whereas the previous implementation was OK IMO.

File: gravitee-am-management-api/gravitee-am-management-api-standalone/gravitee-am-management-api-standalone-distribution/src/main/resources/config/gravitee.yml [hunk 5]
```diff
@@ -281,6 +281,8 @@ email:
   from: noreply@my.domain
   username: user@my.domain
   password: password
+  allowlist:
```
Comment (leleueri) [inline gravitee-am-management-api/gravitee-am-management-api-standalone/gravitee-am-management-api-standalone-distribution/src/main/resources/config/gravitee.yml:284]: Not sure about the name, it is not obvious that this entry refer to the allowed `from` values.

If better name can't be found, maybe a comment maybe useful.

File: gravitee-am-service/src/main/java/io/gravitee/am/service/validators/flow/policy/SendEmailPolicyValidator.java (hunk 2 above)
Comment (leleueri) [reply gravitee-am-service/src/main/java/io/gravitee/am/service/validators/flow/policy/SendEmailPolicyValidator.java:54]: Agree, should be at least a WARN

File: gravitee-am-management-api/gravitee-am-management-api-standalone/gravitee-am-management-api-standalone-distribution/src/main/resources/config/gravitee.yml (hunk 5 above)
Comment (leleueri) [reply gravitee-am-management-api/gravitee-am-management-api-standalone/gravitee-am-management-api-standalone-distribution/src/main/resources/config/gravitee.yml:284]: Also the helm chart will have to be updated to take into account this new setting

File: gravitee-am-service/src/test/java/io/gravitee/am/service/validators/EmailDomainValidatorTest.java [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+        assertEquals(expectedResult, validator.validate(email));
+    }
+
+    private static Stream<Arguments> params_that_must_validate_email(){
+        return Stream.of(
+                Arguments.of("test@example.com", true),
+                Arguments.of("another@mail.com", true),
+                Arguments.of("invalid.email", false),
+                Arguments.of("", false),
+                Arguments.of(null, false)
```
Comment (leleueri) [inline gravitee-am-service/src/test/java/io/gravitee/am/service/validators/EmailDomainValidatorTest.java:68]: Can we add another test with  `evil@mail.com` as all invalid cases are reallying on a non valid email address ?

File: gravitee-am-gateway/gravitee-am-gateway-standalone/gravitee-am-gateway-standalone-distribution/src/main/resources/config/gravitee.yml [hunk 7]
```diff
@@ -220,6 +220,8 @@ email:
   from: noreply@my.domain
   username: user@my.domain
   password: password
+  allowlist:
+    - "*@*.*"
```
Comment (leleueri) [inline gravitee-am-gateway/gravitee-am-gateway-standalone/gravitee-am-gateway-standalone-distribution/src/main/resources/config/gravitee.yml:224]: see comments on the [gravitee.yaml](https://github.com/gravitee-io/gravitee-access-management/pull/2849/files#diff-12f73c7e177a2eda64da20512fd027158e798e496fabb691acf7aa8adb302b40) in the management-api service

**Review by leleueri — CHANGES_REQUESTED** [review]
In addition of all the comments, here is two general ones.

First I have a doubt about the way the error is propagated to the REST Api.
As far as I undestand, we just inform that the flow configuration is invalid but we do not specify that it is due to the email settings. Maybe something to improve here.

Second, IMO this PR is not complete as we do not perform control on the **from** value defined in custom email templates (domain & application level)

File: gravitee-am-service/src/main/java/io/gravitee/am/service/validators/email/EmailDomainValidatorImpl.java [hunk 8]
```diff
@@ -0,0 +1,67 @@
...
+
+    private static List<Pattern> initializeAllowList(ConfigurableEnvironment environment) {
+        List<Pattern> allowList = new ArrayList<>();
+        for (int i = 0; true; i++) {
+            var propertyKey = String.format(EMAIL_ALLOW_LIST, i);
+            var value = environment.getProperty(propertyKey, String.class);
+            if (isNull(value)) {
+                break;
+            }
+            allowList.add(Pattern.compile(WildcardUtils.toRegex(value)));
```
[the remark this answers is not in the document]
Comment (remisultan) [reply gravitee-am-service/src/main/java/io/gravitee/am/service/validators/email/EmailDomainValidatorImpl.java:59]: Usually if it fails it we just let it crash

File: gravitee-am-service/src/main/java/io/gravitee/am/service/validators/flow/policy/SendEmailPolicyValidator.java (hunk 2 above)
Comment (remisultan) [reply gravitee-am-service/src/main/java/io/gravitee/am/service/validators/flow/policy/SendEmailPolicyValidator.java:54]: Let's go with this then

File: gravitee-am-management-api/gravitee-am-management-api-service/src/main/java/io/gravitee/am/management/service/impl/EmailServiceImpl.java (hunk 4 above)
Comment (remisultan) [reply gravitee-am-management-api/gravitee-am-management-api-service/src/main/java/io/gravitee/am/management/service/impl/EmailServiceImpl.java:141]: I implemented something there but removed it after. I'll rollback to thoses changes

File: gravitee-am-management-api/gravitee-am-management-api-standalone/gravitee-am-management-api-standalone-distribution/src/main/resources/config/gravitee.yml (hunk 5 above)
Comment (remisultan) [reply gravitee-am-management-api/gravitee-am-management-api-standalone/gravitee-am-management-api-standalone-distribution/src/main/resources/config/gravitee.yml:284]: It's true we have the helm charts in the repo now 👍 
Good catch thanks

File: gravitee-am-service/src/test/java/io/gravitee/am/service/validators/FlowValidatorImplTest.java (hunk 3 above)
Comment (remisultan) [reply gravitee-am-service/src/test/java/io/gravitee/am/service/validators/FlowValidatorImplTest.java:69]: Will do

**Review by mariano-fernandez — APPROVED** [review]

**Pull request merged**